File: main/binary-tree-postorder-traversal/binary-tree-postorder-traversal.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    private enum Step {
        Display,
        GoLeft,
        GoRight,
    }

    private class State {
        public readonly TreeNode Node;
        public Step Next;

        public State(TreeNode node, Step next) {
            Node = node;
            Next = next;
        }
    }

    public IList<int> PostorderTraversal(TreeNode root) {
        var path = new List<int>();
        if (root == null) return path;

        var stack = new Stack<State>();
        stack.Push(new State(root, Step.GoLeft));

        while (stack.Count != 0) {
            var state = stack.Peek();

            switch (state.Next) {
            case Step.GoLeft:
                state.Next = Step.GoRight;
                var left = state.Node.left;
                if (left != null) stack.Push(new State(left, Step.GoLeft));
                break;

            case Step.GoRight:
                state.Next = Step.Display;
                var right = state.Node.right;
                if (right != null) stack.Push(new State(right, Step.GoLeft));
                break;

            case Step.Display:
                path.Add(state.Node.val);
                stack.Pop();
                break;
            }
        }

        return path;
    }
}

File: main/binary-tree-postorder-traversal/binary-tree-postorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

from enum import Enum, unique

@unique
class Act(Enum):
    DISPLAY = 0
    GO_LEFT = 1
    GO_RIGHT = 2

class State(object):
    def __init__(self, node, act):
        self.node = node
        self.act = act

class Solution:
    def postorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        path = []
        if root is None:
            return path

        stack = [State(root, Act.GO_LEFT)]
        while stack:
            state = stack[-1]

            if state.act == Act.GO_LEFT:
                state.act = Act.GO_RIGHT
                left = state.node.left
                if left is not None:
                    stack.append(State(left, Act.GO_LEFT))
            elif state.act == Act.GO_RIGHT:
                state.act = Act.DISPLAY
                right = state.node.right
                if right is not None:
                    stack.append(State(right, Act.GO_LEFT))
            elif state.act == Act.DISPLAY:
                path.append(state.node.val)
                del stack[-1]
            else:
                raise ValueError('invalid state')

        return path
